doc/developer/design/20240522_mir_representation_types.md

```diff
@@ -0,0 +1,133 @@
+
+This design doc proposes _representation types_---a new notion of type for MIR terms.
+A representation type maps to a subset of constructors/variants of the `Datum` enum.
+
+Making this change should help address issues in optimization, particularly index selection (see ["The Problem"](#the-problem)).
+It will also simplify and clarify what MIR is doing.
+
+## The Problem
+
+[Postgres has several types for character data:](https://www.postgresql.org/docs/current/datatype-character.html)
+
+- `char(n)`, `bpchar(n)` for fixed-length strings with a fixed-length representation (whitespace padded)
+- `varchar(n)` for fixed-length strings with a variable-length representation
+- `varchar` for arbitrary-length strings with a variable-length representation
+- `bpchar` for artbirary-length strings with a variable-length representation where trailing whitespace is trimmed
+- `text` for arbitrary-length strings with a variable-length representation
+
+(There are also `"char"` and `name`, but they are very much internal to Postgres---let's ignore them for the moment.)
+
+Materialize implements all five of these character sequence types using `Datum::String`.
+Materialize only implements meaningful operations on the `text` type; to operate on other character sequence types, one must convert to `text`.
+Conversion to `text` is a no-op at runtime, since we already have the string.
+We freely insert implicit coercions to `text` in the early stages of query processing.
+For correct error behavior, however, we need to carefully keep track of failures to meet length limits.
+
+These implicit coercions---often of the form `varchar_to_text(#c)` for some column number `c`---get in the way during join planning.
+Even if we have an index on column `c` as a `varchar`, we cannot use that index to plan a join on `varchar_to_text(#c)`: not only are the types different, the terms are different, too.
+We know these implicit coercions interfere with joins; they may also interfere with other aspects of optimization.
+
+## Success Criteria
+
+1. Plan joins without implicit coercions between equivalent representation types getting in the way.
+2. It should be possible for us to elide noop casts at the MIR level.
+3. The catalog should reflect SQL-level types.
+4. There is clear guidance for customers regarding the different character sequence types.
+5. Notices help customers avoid pitfalls with character sequence types.
+
+## Out of Scope
+
+We will not rethink other components of MIR types, like nullability or unique keys.
+
+## Solution Proposal
+
+Having MIR be typechecked in terms of representation types demands a few changes, which can be broken into three PRs:
+
+1. Rename the existing `ScalarType` to `SqlScalarType`. (We may want to rename `ColumnType` and `RelationType`, as well.)
+2. Have MIR work with `ReprScalarType` (see [MVP](#minimal-viable-prototype)).
```

Comment:
I don't love that we have two parallel naming schemes: "HIR", "MIR", and "LIR" for expressions, and "Sql" and "Repr" for types. Is it right that HIR types are always `SqlScalarTypes` and MIR and LIR types are always `ReprScalarTypes`? Is there some intuition that suggests we'll only need the two ends of the spectrum for types, even though we needed three IRs for expressions? Or do we think it's likely we'll one day have `HirScalarType`, `MirScalarType`, and `LirScalarType`?

Reply:
I think we might also call them `HirScalarType` and `MirScalarType`. LIR already has a precedent of using `MirScalarExpr` for its scalars. If we want to have the IR families aligned in their prefixes we might introduce type aliases

```rust
type LirScalarType = MirScalarType;
type LirScalarExpr = MirScalarExpr;
```

Reply:
I'm easy on names. Right now LIR is effectively untyped. If we were to add types to LIR, they would be the same as MIR types. (@frankmcsherry, @antiguru, and I have discussed the idea of a bytecode/instruction LIR, which might end up having a finer grained type system.)